File: hash-map/contains-duplicate-2/main.go
package main

func containsNearbyDuplicate(nums []int, k int) bool {
	myMap := map[int][]int{}

	for idx, num := range nums {
		indices := myMap[num]
		if len(indices) > 0 {
			if idx-indices[len(indices)-1] <= k {
				return true
			} else {
				myMap[num] = append(indices, idx)
			}
		} else {
			myMap[num] = []int{idx}
		}
	}

	return false
}

func main() {
	containsNearbyDuplicate([]int{1, 0, 1, 1}, 1)
}

File: hash-map/contains-duplicate-2/main.py
from typing import List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        myMap = {}
        for idx in range(len(nums)):
            num = nums[idx]
            if num in myMap:
                if idx - myMap[num][len(myMap[num]) - 1] <= k:
                    return True
                else:
                    myMap[num].append(idx)
            else:
                myMap[num] = [idx]
        return False
